lib/auth/predicate/predicate.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicate
+
+import (
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+	"github.com/vulcand/predicate"
+)
+
+// NamedParameter is an object with a name that can be added to the environment in which a predicate expression runs.
+type NamedParameter interface {
+	// GetName returns the name of the object.
+	GetName() string
+}
+
+func buildEnv(objects ...NamedParameter) map[string]any {
+	env := make(map[string]any)
+	for _, obj := range objects {
+		env[obj.GetName()] = obj
+	}
+
+	return env
+}
+
+func newParser(env map[string]any) (predicate.Parser, error) {
+	getIdentifierInEnv := func(selector []string) (any, error) {
+		return getIdentifier(env, selector)
+	}
+
+	return predicate.NewParser(predicate.Def{
+		Operators: predicate.Operators{
+			AND: builtinOpAnd,
+			OR:  builtinOpOr,
+			NOT: builtinOpNot,
+			EQ:  builtinOpEquals,
+			LT:  builtinOpLT,
+			GT:  builtinOpGT,
+			LE:  builtinOpLE,
+			GE:  builtinOpGE,
+		},
+		Functions: map[string]any{
+			"add":            builtinAdd,
+			"sub":            builtinSub,
+			"mul":            builtinMul,
+			"div":            builtinDiv,
+			"xor":            builtinXor,
+			"split":          builtinSplit,
+			"upper":          builtinUpper,
+			"lower":          builtinLower,
+			"contains":       builtinContains,
+			"first":          builtinFirst,
+			"append":         builtinAppend,
+			"array":          builtinArray,
+			"replace":        builtinReplace,
+			"len":            builtinLen,
+			"regex":          builtinRegex,
+			"matches":        builtinMatches,
+			"contains_regex": builtinContainsRegex,
+			"map_insert":     builtinMapInsert,
+			"map_remove":     builtinMapRemove,
+		},
+		GetIdentifier: getIdentifierInEnv,
+		GetProperty:   getProperty,
+	})
+}
+
+// PredicateAccessChecker checks access/permissions to access certain resources by evaluating AccessPolicy resources.
+type PredicateAccessChecker struct {
+	polices []types.Policy
+}
+
+// NewPredicateAccessChecker creates a new PredicateAccessChecker with a set of policies describing the permissions.
+func NewPredicateAccessChecker(policies []types.Policy) *PredicateAccessChecker {
+	return &PredicateAccessChecker{policies}
+}
+
+// CheckAccessToNode checks if a given user has access to a Server Access node.
+func (c *PredicateAccessChecker) CheckAccessToNode(node *Node, user *User) (bool, error) {
```

Comment:
I want to know how these access checks will be merged with our existing RBAC. With they both run or just one or the other? What if roles say no access, policies say access is allowed? Or vice-versa? I'm hoping resource access requests will be respected here

Reply:
The idea is that we run both. We deny by default and grant a request if one of the system incl resource access requests responds true, for full backwards compatibility. If any system returns with an explicit deny rule, this overrules any eventual decision made by the other system.

Reply:
I'm not sure how you will support this with just `CheckAccessToNode`, would you need like a `CheckNodeDeny` and `CheckNodeAllow` or something? Is that coming later on?

And for existing RBAC it is much more generic so we don't need specific `CheckAccessToNode`, `CheckAccessToDatabase`, etc... what's the plan for adding more access checkable types, make the interface more generic or add more functions?

Reply:
I might port that over to this PR, but in #18020 I realized we need three-way access decisions (allow, undecided, explicit deny) to allow those to be combined and have made those changes on top of this PR.

Reply:
I think it makes sense to merge this as is for now to avoid complex and history-mangling merge operations to port that 3-way over to this PR. Can you take another look? @nklaassen @fspmarshall